### PR TITLE
Fixing groups API validation

### PR DESCRIPTION
**Pull request by fao89**
https://pulp.plan.io/issues/7329
closes #7329

Please be sure you have read our documentation on creating PRs:
https://docs.pulpproject.org/contributing/pull-request-walkthrough.html

### CODE REVIEW

File: pulpcore/app/viewsets/user.py [hunk 1]
```diff
@@ -98,7 +100,12 @@ def get_model_permission(self, request):
 
             data[key] = value
 
-        return get_object_or_404(Permission, **data)
+        try:
+            permission = Permission.objects.get(**data)
+        except (Permission.MultipleObjectsReturned, Permission.DoesNotExist, FieldError) as exc:
+            raise ValidationError(_(str(exc)))
```
Comment (mdellweg) [inline pulpcore/app/viewsets/user.py:106]: I do not believe, gettext works in this situation. How is it supposed to extract the translatable string here?

File: pulpcore/app/viewsets/user.py [hunk 2]
```diff
@@ -263,7 +293,11 @@ def create(self, request, group_pk):
             raise ValidationError(
                 _("Please provide one of the following values for User: 'pk', 'id', 'username'")
             )
-        user = get_object_or_404(get_user_model(), **request.data)
+        User = get_user_model()
```
Comment (mdellweg) [inline pulpcore/app/viewsets/user.py:296]: Would it be OK do make this call at the top level? Or is django insufficiently initialized at that point?

File: pulpcore/app/viewsets/user.py [hunk 3]
```diff
@@ -263,7 +293,11 @@ def create(self, request, group_pk):
...
+        try:
+            user = User.objects.get(**request.data)
+        except (User.MultipleObjectsReturned, User.DoesNotExist, FieldError) as exc:
```
Comment (mdellweg) [inline pulpcore/app/viewsets/user.py:299]: How can `MultipleObjectsReturned` happen, if we require to provide `pk`, `id` or `username`?
Is this overly cautious?

Comment (fao89) [reply pulpcore/app/viewsets/user.py:299]: it was a bad copy/paste

File: pulpcore/app/viewsets/user.py (hunk 1 above)
Comment (fao89) [reply pulpcore/app/viewsets/user.py:106]: I do not know how gettext works, `str(exc)` returns a string I imagined it would work

Comment (mdellweg) [reply pulpcore/app/viewsets/user.py:108]: Well, I thought it must be able to extract the strings without calling the code. So it probably looks for calls to ’_("...")’.

Comment (dkliban) [reply pulpcore/app/viewsets/user.py:108]: The exception text should already be translated if a translation is available. Please remove the _() around str(exc).

File: pulpcore/app/viewsets/user.py [hunk 4]
```diff
@@ -164,10 +173,33 @@ class GroupObjectPermissionViewSet(NamedModelViewSet):
...
+    def get_object_pk(self, request):
+        """Get object pk."""
+
+        if "obj" not in request.data:
+            raise ValidationError(_("Please provide 'obj' value"))
+        try:
+            obj_pk = request.data["obj"].strip("/").split("/")[-1]
+            uuid.UUID(obj_pk)
+        except (AttributeError, ValueError):
+            raise ValidationError(_(f"Invalid value for 'obj': {request.data['obj']}"))
```
Comment (mdellweg) [inline pulpcore/app/viewsets/user.py:185]: Stackoverflow says nope:
https://stackoverflow.com/questions/49797658/how-to-use-gettext-with-python-3-6-f-strings

```suggestion
            raise ValidationError(_("Invalid value for 'obj': " + str(request.data['obj'])))
```
Or the good old "format" of course (Usually this helps people translating if they see the whole string template).
```suggestion
            raise ValidationError(_("Invalid value for 'obj': {obj}").format(request.data['obj']))
```

File: pulpcore/app/viewsets/user.py [hunk 5]
```diff
@@ -164,10 +173,33 @@ class GroupObjectPermissionViewSet(NamedModelViewSet):
...
+        if "permission" not in request.data:
+            raise ValidationError(_("Please provide 'permission' value"))
+
         codename = request.data["permission"].split(".")[-1]
-        return get_object_or_404(Permission, codename=codename)
+
+        try:
+            permission = Permission.objects.get(codename=codename)
+        except (Permission.MultipleObjectsReturned, Permission.DoesNotExist, FieldError) as exc:
+            raise ValidationError(_(str(exc)))
```
Comment (mdellweg) [inline pulpcore/app/viewsets/user.py:200]:
```suggestion
            raise ValidationError(str(exc))
```
Won't work at all, but the exception should only use translatable strings anyway.

File: pulpcore/app/viewsets/user.py [hunk 6]
```diff
@@ -263,7 +295,10 @@ def create(self, request, group_pk):
             raise ValidationError(
                 _("Please provide one of the following values for User: 'pk', 'id', 'username'")
             )
-        user = get_object_or_404(get_user_model(), **request.data)
+        try:
+            user = User.objects.get(**request.data)
+        except (User.DoesNotExist, FieldError) as exc:
+            raise ValidationError(_(str(exc)))
```
Comment (mdellweg) [inline pulpcore/app/viewsets/user.py:301]: same here
```suggestion
            raise ValidationError(str(exc))
```

**Review by mdellweg — CHANGES_REQUESTED** [review]
Not sure if i'm the right one to review.
But at least with the gettext stuff i am pretty confident now.

**Review by mdellweg — APPROVED** [review]

File: pulpcore/app/viewsets/user.py [hunk 7]
```diff
@@ -274,7 +309,7 @@ def destroy(self, request, group_pk, pk):
         Remove a user from a group.
         """
         group = Group.objects.get(pk=group_pk)
-        user = get_object_or_404(get_user_model(), pk=pk)
+        user = get_object_or_404(User, pk=pk)
```
Comment (dkliban) [inline pulpcore/app/viewsets/user.py:312]: Is this going to be a problem for galaxy_ng?

Comment (fao89) [reply pulpcore/app/viewsets/user.py:312]: nope, at the beginning of this file (line 22) I declared:
```python 
User = get_user_model()
```

**Pull request merged**